implementation/src/main/java/io/smallrye/config/KeyMap.java

```diff
@@ -0,0 +1,280 @@
+            return this;
+        }
+        String seg = iter.next();
+        KeyMap<V> next = seg.equals("*") ? getOrCreateAny() : computeIfAbsent(seg, k -> new KeyMap<>());
+        return next.findOrAdd(iter);
+    }
+
+    public KeyMap<V> findOrAdd(final Iterable<String> i) {
+        return findOrAdd(i.iterator());
+    }
+
+    public KeyMap<V> findOrAdd(final String... keys) {
+        return findOrAdd(keys, 0, keys.length);
+    }
+
+    public KeyMap<V> findOrAdd(final String[] keys, int off, int len) {
+        String seg = keys[off];
+        KeyMap<V> next = seg.equals("*") ? getOrCreateAny() : computeIfAbsent(seg, k -> new KeyMap<>());
+        return off + 1 > len - 1 ? next : next.findOrAdd(keys, off + 1, len);
+    }
+
+    public V findRootValue(final String path) {
+        return findRootValue(new NameIterator(path));
+    }
+
+    public V findRootValue(final NameIterator ni) {
+        KeyMap<V> result = find(ni);
+        return result == null ? null : result.getRootValue();
+    }
+
+    public boolean hasRootValue(final String path) {
+        return hasRootValue(new NameIterator(path));
+    }
+
+    public boolean hasRootValue(final NameIterator ni) {
+        KeyMap<V> result = find(ni);
+        return result != null && result.hasRootValue();
+    }
+
+    public <P, V2> KeyMap<V2> map(P param, BiFunction<P, V, V2> conversion) {
+        return map(param, conversion, new IdentityHashMap<>());
+    }
+
+    public <V2> KeyMap<V2> map(Function<V, V2> conversion) {
+        return map(conversion, Functions.functionBiFunction());
+    }
+
+    <P, V2> KeyMap<V2> map(P param, BiFunction<P, V, V2> conversion, IdentityHashMap<KeyMap<V>, KeyMap<V2>> cached) {
+        if (cached.containsKey(this)) {
+            return cached.get(this);
+        }
+        KeyMap<V2> newMap = new KeyMap<>(size());
+        cached.put(this, newMap);
+        Set<Entry<String, KeyMap<V>>> entries = entrySet();
+        for (Entry<String, KeyMap<V>> entry : entries) {
+            newMap.put(entry.getKey(), entry.getValue().map(param, conversion, cached));
+        }
+        KeyMap<V> any = getAny();
+        if (any != null) {
+            newMap.putAny(any.map(param, conversion, cached));
+        }
+        if (hasRootValue()) {
+            newMap.putRootValue(conversion.apply(param, getRootValue()));
+        }
+        return newMap;
+    }
+
+    public StringBuilder toString(StringBuilder b) {
+        b.append("KeyMap(");
+        V rootValue = this.rootValue;
+        if (rootValue == NO_VALUE) {
+            b.append("no value");
+        } else {
+            b.append("value=").append(rootValue);
+        }
+        b.append(") {");
+        final Iterator<Entry<String, KeyMap<V>>> iterator = entrySet().iterator();
+        KeyMap<V> any = this.any;
+        if (iterator.hasNext()) {
+            Entry<String, KeyMap<V>> entry = iterator.next();
+            b.append(entry.getKey()).append("=>").append(entry.getValue());
```

Comment:
This is needlessly expensive.  It would be significantly better (faster and less allocation-intensive) to append the `"=>"`, and then recursively call the value's `toString(StringBuilder)` method.